adjust subgraph to Hats-Protocol v1-beta.1 version

gershido opened the pull request:
The PR includes:
- Support for tree linking: 
The `Tree` object now includes the fields `linkedToTree`, `linkedToHat` and `linkedFromTrees`. The `Hat` object includes the field `linkedTrees` for all trees linked to this hat.
- Records for bad standing: 
The `Hat` object records all the wearers with bad standing in the field `badStandings`.
- Adjustments for the new hat Id structure.

gershido committed:
adjust subgraph to version v1-beta.1

gershido committed:
add prettier

spengrah commented on README.md:
```diff
@@ -24,29 +24,33 @@ Get a specific tree:
 {
-  hat(id: "0x0000000101000000000000000000000000000000000000000000000000000000") {
+  hat(
+    id: "0x0000000101000000000000000000000000000000000000000000000000000000"
```
I think this should be `0x0000000100010000000000000000000000000000000000000000000000000000` now

spengrah replied:
And same for subsequent example ids below

spengrah commented on README.md:
```diff
@@ -24,29 +24,33 @@ Get a specific tree:
 {
-  hat(id: "0x0000000101000000000000000000000000000000000000000000000000000000") {
+  hat(
+    id: "0x0000000101000000000000000000000000000000000000000000000000000000"
```
I think this (and example ids below) should now be `0x0000000100010000000000000000000000000000000000000000000000000000`

spengrah replied:
Realizing that you didn't make any actual changes to the readme (diffs are just thanks to prettier), so you can ignore this one for now

spengrah commented on src/hats.ts:
```diff
@@ -108,13 +118,15 @@ export function handleHatEligibilityChanged(
   hatEligibilityChangedEvent.blockNumber = event.block.number.toI32();
   hatEligibilityChangedEvent.timestamp = event.block.timestamp;
   hatEligibilityChangedEvent.transactionID = event.transaction.hash;
-  hatEligibilityChangedEvent.hatId = hatIdToHex(event.params.hatId);
```
you're removing this because its duplicative with `.hat`, correct?

gershido replied:
yes

spengrah commented on src/hats.ts:
```diff
@@ -211,18 +228,97 @@ export function handleHatToggleChanged(event: HatToggleChanged): void {
+  topHatLinkRequestedEvent.save();
+}
+
+export function handleTopHatLinked(event: TopHatLinked): void {
```
Would it make sense here to update `hat.admin` and `hat.level` for the newly linked tophat? If it were costless to do so, my sense is Yes since that would bring those datapoints in sync with the protocol state.

spengrah replied:
Similarly, but also very differently, how do you think we should handle `hat.level` for all the hats within the newly linked tree (and any trees linked to it)? 

Technically, their levels are now all changed, since `Hats.getHatLevel()` recurses through linked trees to derive the hat's level. But I'm guessing it would be quite expensive to update the data for all those hats.

Maybe some kind of lazy updating, such as if/when any of those hats are updated in some other way?

spengrah commented on schema.graphql:
```diff
@@ -28,16 +30,19 @@ type Wearer @entity {
+  linkedToTree: Tree
+  linkedToHat: Hat
   hats: [Hat!]! @derivedFrom(field: "tree")
+  linkedFromTrees: [Tree!]! @derivedFrom(field: "linkedToTree")
```
I'm finding the "linkedTo" and "linkedFrom" labels a bit confusing. Maybe its just me being dense and I'll get used to it, but I'm wondering if "parent / child" semantics might be clearer. I'm not sold on this, but perhaps something like "childOfTree" instead of "linkedToTree", and "parentOfTrees" instead of "linkedFromTrees"? 

I could be getting those backwards, btw -- which I suppose is kind of my point 😉

spengrah reviewed (changes requested):
This is great work @gershido! Really impressed by and grateful for your speed and dedication 🙏 

I left a few comments and questions in-line, the biggest of which is related to how we should handle changes to properties (esp `level`) of hats in trees that are linked. Marked this as "requesting changes" as a result.

gershido committed:
fix bugs and update readme

gershido commented:
Thank you for the review!
Pushed the following changes:
- Changing the admin of the top hat whenever a tree is linked.
- Changed the `level` field in the `Hat` entity to `levelAtLocalTree` to represent the local level. The only solution I see now to the level issue is to include a new field `levelAtGlobalTree` to the `Tree` entity to represent the level of its top hat in the global tree (in case it was linked). Then, if tree X is linked, we'll update this field in X and in every tree that is linked to X recursively. To get the level of a hat you'll need to add `levelAtLocalTree` with `levelAtGlobalTree`. The problem is that updating `levelAtGlobalTree` theoretically has unbounded cost because there's no limit to the length of the possible chain of trees linked to one another. 
- Updated the naming in the `Tree` entity as suggested, to `childOfTree` and `parentOfTrees`
- Updated the README

spengrah commented:
> * Changed the `level` field in the `Hat` entity to `levelAtLocalTree` to represent the local level. The only solution I see now to the level issue is to include a new field `levelAtGlobalTree` to the `Tree` entity to represent the level of its top hat in the global tree (in case it was linked). Then, if tree X is linked, we'll update this field in X and in every tree that is linked to X recursively. To get the level of a hat you'll need to add `levelAtLocalTree` with `levelAtGlobalTree`. The problem is that updating `levelAtGlobalTree` theoretically has unbounded cost because there's no limit to the length of the possible chain of trees linked to one another.

I agree with your assessment of the problem, and I really like your `levelAtLocalTree`-only approach. Nice and simple, and IMO its totally fine to have front-ends calculate the `levelAtGlobalTree` for themselves as needed based on other indexed data.

spengrah commented on README.md:
```diff
@@ -2,6 +2,10 @@
 
+# Important Note:
+
+The `status` and `badStandings` fields of a `Hat` entity, track the status and standing as currently recorded in the Hats contract. It is NOT tracking the eligibility and toggle modules in case they exist and thus only represent the last recorded state in the Hats contract. In case they depend on the state of external contracts, these fields in the subgraph may be out of sync with the onchain sta ...
```
nice add -- important note!

spengrah commented on README.md:
````diff
@@ -127,6 +170,36 @@ The following query will get the wearers of the hat:
 }
 ```
 
+The following query will get the all the wearers with a bad standing:
````
including all these sample queries is awesome!

spengrah commented on package.json:
```diff
@@ -4,7 +4,7 @@
     "codegen": "graph codegen",
     "build": "graph build",
-    "deploy": "graph deploy --node https://api.thegraph.com/deploy/ hats-protocol/hats-protocol-goerli",
+    "deploy": "graph deploy --node https://api.thegraph.com/deploy/ gershido/hats-protocol-goerli",
```
is this correct?

gershido replied:
No, thanks! Used my account for testing

spengrah commented on schema.graphql:
```diff
@@ -9,7 +9,7 @@ type Hat @entity {
   mutable: Boolean!
   imageUri: String!
-  level: Int!
+  levelAtLocalTree: Int!
```
this is absolutely the right solution 🫡

spengrah commented on schema.graphql:
```diff
@@ -30,10 +30,10 @@ type Wearer @entity {
   "Tree ID is its top hat domain - first 4 bytes of the top hat ID"
   id: ID!
-  linkedToTree: Tree
+  childOfTree: Tree
```
much clearer now :)

spengrah reviewed (commented):
Thanks for making these changes -- it's looking really good!

I left just one small question about the package.json diff. Otherwise, good to go!

gershido committed:
fix deploy command in package.json
